From: Jeremy Bicha <jbicha@debian.org>
Date: Sat, 9 Feb 2019 16:32:03 -0500
Subject: sound: Allow volume to be set above 100%.

Some systems have low maximum volume set (like x220), allow, from an option
in regolith-control-center to set it above that 100% limit from g-s-d
(keyboard) and gnome-shell.
Only show the above 100% volume option if:
1. you are in an ubuntu session
2. the selected output supports amplified volume. If so:
   present the settings to google that on and off. It will enable
   GNOME Shell and media keys to set the sound above 100%. If not
   enabled, volume and sliders are all capped to 100%. (LP: #1706524)
Modified from original patch in unity-control-center from Lars Uebernickel.
Adapted for Ubuntu regolith-control-center by Didier Roche
Updated for GNOME 3.32 by Jeremy Bicha
Origin: ubuntu
Bug-Ubuntu: https://launchpad.net/bugs/1706524
Bug: https://bugzilla.gnome.org/show_bug.cgi?id=710424
---
 panels/sound/cc-sound-panel.c  | 20 ++++++++++++++++++++
 panels/sound/cc-sound-panel.ui | 42 ++++++++++++++++++++++++++++++++++++++++++
 2 files changed, 62 insertions(+)

diff --git a/panels/sound/cc-sound-panel.c b/panels/sound/cc-sound-panel.c
index 5e6e03d..b1b5ec9 100644
--- a/panels/sound/cc-sound-panel.c
+++ b/panels/sound/cc-sound-panel.c
@@ -65,6 +65,9 @@ struct _CcSoundPanel
   CcProfileComboBox *output_profile_combo_box;
   GtkListBoxRow     *output_profile_row;
   CcVolumeSlider    *output_volume_slider;
+  CcVolumeSlider    *output_volume_slider_ubuntu;
+  GtkWidget         *allow_amplify_switch;
+  GtkListBox        *ubuntu_output_listbox;
   CcStreamListBox   *stream_list_box;
   GtkListBoxRow     *subwoofer_row;
   CcSubwooferSlider *subwoofer_slider;
@@ -88,6 +91,8 @@ allow_amplified_changed_cb (CcSoundPanel *self)
 {
   cc_volume_slider_set_is_amplified (self->output_volume_slider,
                                      g_settings_get_boolean (self->sound_settings, "allow-volume-above-100-percent"));
+  cc_volume_slider_set_is_amplified (self->output_volume_slider_ubuntu,
+                                     g_settings_get_boolean (self->sound_settings, "allow-volume-above-100-percent"));
 }
 
 static void
@@ -98,6 +103,7 @@ set_output_stream (CcSoundPanel   *self,
   gboolean can_fade = FALSE, has_lfe = FALSE;
 
   cc_volume_slider_set_stream (self->output_volume_slider, stream, CC_STREAM_TYPE_OUTPUT);
+  cc_volume_slider_set_stream (self->output_volume_slider_ubuntu, stream, CC_STREAM_TYPE_OUTPUT);
   cc_level_bar_set_stream (self->output_level_bar, stream, CC_STREAM_TYPE_OUTPUT);
 
   if (stream != NULL)
@@ -259,6 +265,9 @@ cc_sound_panel_class_init (CcSoundPanelClass *klass)
   gtk_widget_class_bind_template_child (widget_class, CcSoundPanel, output_profile_combo_box);
   gtk_widget_class_bind_template_child (widget_class, CcSoundPanel, output_profile_row);
   gtk_widget_class_bind_template_child (widget_class, CcSoundPanel, output_volume_slider);
+  gtk_widget_class_bind_template_child (widget_class, CcSoundPanel, output_volume_slider_ubuntu);
+  gtk_widget_class_bind_template_child (widget_class, CcSoundPanel, allow_amplify_switch);
+  gtk_widget_class_bind_template_child (widget_class, CcSoundPanel, ubuntu_output_listbox);
   gtk_widget_class_bind_template_child (widget_class, CcSoundPanel, stream_list_box);
   gtk_widget_class_bind_template_child (widget_class, CcSoundPanel, subwoofer_row);
   gtk_widget_class_bind_template_child (widget_class, CcSoundPanel, subwoofer_slider);
@@ -293,12 +302,23 @@ cc_sound_panel_init (CcSoundPanel *self)
                            G_CONNECT_SWAPPED);
   allow_amplified_changed_cb (self);
 
+  if (strstr (g_getenv("XDG_CURRENT_DESKTOP"), "ubuntu") != NULL) {
+    gtk_widget_set_visible(self->ubuntu_output_listbox, TRUE);
+    gtk_widget_set_visible(GTK_WIDGET (self->output_volume_slider), FALSE);
+    g_settings_bind (self->sound_settings, "allow-volume-above-100-percent",
+                     self->allow_amplify_switch, "active", G_SETTINGS_BIND_DEFAULT);
+  } else {
+    gtk_widget_set_visible(self->ubuntu_output_listbox, FALSE);
+    gtk_widget_set_visible(GTK_WIDGET (self->output_volume_slider), TRUE);
+  }
+
   self->mixer_control = gvc_mixer_control_new ("GNOME Settings");
   gvc_mixer_control_open (self->mixer_control);
 
   cc_stream_list_box_set_mixer_control (self->stream_list_box, self->mixer_control);
   cc_volume_slider_set_mixer_control (self->input_volume_slider, self->mixer_control);
   cc_volume_slider_set_mixer_control (self->output_volume_slider, self->mixer_control);
+  cc_volume_slider_set_mixer_control (self->output_volume_slider_ubuntu, self->mixer_control);
   cc_subwoofer_slider_set_mixer_control (self->subwoofer_slider, self->mixer_control);
   cc_device_combo_box_set_mixer_control (self->input_device_combo_box, self->mixer_control, FALSE);
   cc_device_combo_box_set_mixer_control (self->output_device_combo_box, self->mixer_control, TRUE);
diff --git a/panels/sound/cc-sound-panel.ui b/panels/sound/cc-sound-panel.ui
index b5706e2..3eb3003 100644
--- a/panels/sound/cc-sound-panel.ui
+++ b/panels/sound/cc-sound-panel.ui
@@ -35,6 +35,48 @@
                     <property name="visible">True</property>
                   </object>
                 </child>
+                <!-- Over-Amplification -->
+                <child>
+                  <object class="GtkListBox" id="ubuntu_output_listbox">
+                    <property name="visible">False</property>
+                    <property name="selection_mode">none</property>
+                    <child>
+                      <object class="HdyPreferencesRow">
+                        <property name="visible">true</property>
+                        <property name="activatable">0</property>
+                        <child>
+                          <object class="CcVolumeSlider" id="output_volume_slider_ubuntu">
+                            <property name="visible">True</property>
+                            <property name="margin_top">12</property>
+                            <property name="margin_bottom">12</property>
+                            <property name="margin_end">12</property>
+                          </object>
+                        </child>
+                      </object>
+                    </child>
+                    <child>
+                      <object class="HdyActionRow">
+                        <property name="visible">true</property>
+                        <property name="title" translatable="yes">Over-Amplification</property>
+                        <property name="subtitle" translatable="yes">Allows raising the volume above 100%. This can result in a loss of audio quality; it is better to increase application volume settings, if possible.</property>
+                        <property name="subtitle-lines">6</property>
+                        <property name="activatable-widget">allow_amplify_switch</property>
+                        <property name="use-underline">true</property>
+                        <child>
+                          <object class="GtkSwitch" id="allow_amplify_switch">
+                            <property name="visible">true</property>
+                            <property name="halign">end</property>
+                            <property name="valign">center</property>
+                          </object>
+                        </child>
+                      </object>
+                    </child>
+                    <style>
+                      <class name="content"/>
+                    </style>
+                  </object>
+                </child>
+
                 <child>
                   <object class="GtkLabel">
                     <property name="visible">True</property>
